src/inspector/tracing_agent.cc

```diff
@@ -0,0 +1,108 @@
+#include "tracing_agent.h"
+
+#include "env-inl.h"
+#include "v8.h"
+
+#include <set>
+#include <sstream>
+
+namespace node {
+namespace inspector {
+namespace protocol {
+
+namespace {
+using v8::platform::tracing::TraceWriter;
+
+class InspectorTraceWriter : public node::tracing::AsyncTraceWriter {
+ public:
+  explicit InspectorTraceWriter(NodeTracing::Frontend* frontend)
+                                : frontend_(frontend) {}
+
+  void AppendTraceEvent(
+      v8::platform::tracing::TraceObject* trace_event) override {
+    if (!json_writer_)
+      json_writer_.reset(TraceWriter::CreateJSONTraceWriter(stream_, "value"));
+    json_writer_->AppendTraceEvent(trace_event);
+  }
+
+  void Flush() override {
+    if (!json_writer_)
+      return;
+    json_writer_.reset();
+    std::ostringstream result(
+        "{\"method\":\"NodeTracing.dataCollected\",\"data\":",
+        std::ostringstream::ate);
+    result << stream_.str();
+    result << "}";
+    frontend_->sendRawNotification(result.str());
+    stream_.str("");
+  }
+
+  void Flush(bool) override {
+    Flush();
+  }
+
+ private:
+  std::unique_ptr<TraceWriter> json_writer_;
+  std::ostringstream stream_;
+  NodeTracing::Frontend* frontend_;
+};
+}  // namespace
+
+TracingAgent::TracingAgent(Environment* env)
+                           : env_(env),
+                             trace_writer_(
+                                  tracing::Agent::EmptyClientHandle()) {
+}
+
+TracingAgent::~TracingAgent() {
+  trace_writer_.reset();
+}
+
+void TracingAgent::Wire(UberDispatcher* dispatcher) {
+  frontend_ = std::make_unique<NodeTracing::Frontend>(dispatcher->channel());
+  NodeTracing::Dispatcher::wire(dispatcher, this);
+}
+
+DispatchResponse TracingAgent::start(
+    std::unique_ptr<protocol::NodeTracing::TraceConfig> traceConfig) {
+  if (trace_writer_ != nullptr) {
+    return DispatchResponse::Error(
+        "Call NodeTracing::end to stop tracing before updating the config");
+  }
+
+  std::set<std::string> categories_set;
+  protocol::Array<std::string>* categories =
+      traceConfig->getIncludedCategories();
+  for (size_t i = 0; i < categories->length(); i++)
+    categories_set.insert(categories->get(i));
+
+  if (categories_set.empty())
+    return DispatchResponse::Error("At least one category should be enabled");
+
+  trace_writer_ = env_->tracing_agent()->AddClient(
+      categories_set, std::make_unique<InspectorTraceWriter>(frontend_.get()));
+  return DispatchResponse::OK();
+}
+
+DispatchResponse TracingAgent::end() {
+  trace_writer_.reset();
+  return DispatchResponse::OK();
+}
+
+DispatchResponse TracingAgent::getCategories(
+    std::unique_ptr<protocol::Array<String>>* categories) {
+  *categories = Array<String>::create();
+  categories->get()->addItem("node");
```

Comment:
Is it a temp stub implementation? The actual list is dynamic and should be pulled from TracingController. Right?

Reply:
Currently these are adhoc in core. We have macros for defining them, and once we have the js emit function, the category could be somewhat arbitrary. We'll need to make sure we have a good centralized place to define these.

There's one missing: `node.fs.sync'

Reply:
@jasnell added the category.

Reply:
Created https://github.com/nodejs/diagnostics/issues/191 to track this as a follow-on feature.